### ISSUE TITLE
samtools sort both succeeds and fails

### ISSUE THREAD

**Issue by hepcat72** [galaxyproject/galaxy#8717]
I don't understand what is going on here.  This is the second time I've tried running this workflow (a flow that maps CLIP-Seq data).  The last time I tried it, I was getting "unable to finish job" errors from `samtools sort`.  I think I may have mentioned it in gitter, but I don't recall whether that went anywhere.  At the time, according to my notes, I copied the bam files to our cluster and ran the commands manually (and they worked).

Today, I tried a copy of the same workflow and started seeing the errors again.  The path in the flow leading to this error outcome is:

paired collection->cutadapt->RNA STAR->RmDup->Filter BAM or SAM->samtools sort...

All subsequent steps in the flow obtain a paused status.

So I tried gathering all of the `Filter SAM or BAM` outputs (just before the samtools sort error) into a collection and running a new smaller workflow that starts with that as input.  I watched it run and the progress bar was filling up green but there were datasets showing up above it in red with the `Unable to finish job` errors:

![samtools_sort_failure](https://user-images.githubusercontent.com/2300532/65710946-f6e28280-e061-11e9-90da-ae3aa49569f3.png)

Here are the smaller and original workflows:

[workflows.zip](https://github.com/galaxyproject/galaxy/files/3659122/workflows.zip)

So why am I getting these errors and everything works on the command line without error?

**Comment by hepcat72** [comment]
I just tried downloading the smallest bam file and re-uploading it and put it into a collection all its own.  I then ran an even smaller workflow on it that only has the samtools sort step.  The resulting output was a paused item in my history that had the message: `Execution of this dataset's job is paused because its input datasets are in an error state.`

**Comment by hepcat72** [comment]
I spent an hour or two trying to get the smaller galaxy workflow to work.  I eventually gave up, downloaded the files that I was submitting to the smaller workflow, and ran it all on the command line (tcsh):

```
>foreach b ( *.bam )
samtools sort -n "${b}" > "${b}.nsort.bam"
echo $b
end
[bam_sort_core] merging from 4 files and 1 in-memory blocks...
Filter BAM or SAM on data 140.bam
Filter BAM or SAM on data 188.bam
Filter BAM or SAM on data 236.bam
[bam_sort_core] merging from 5 files and 1 in-memory blocks...
Filter BAM or SAM on data 92.bam
>foreach f ( *.nsort.bam )
samtools fixmate "$f" "$f.fixed.bam"
echo $f
end
Filter BAM or SAM on data 140.bam.nsort.bam
Filter BAM or SAM on data 188.bam.nsort.bam
Filter BAM or SAM on data 236.bam.nsort.bam
Filter BAM or SAM on data 92.bam.nsort.bam
>foreach f ( *.fixed.bam )
sambamba view --filter="mapping_quality >= 20 and not (unmapped or mate_is_unmapped)" -f bam -o "$f.filt.bam" "$f"
echo $f
end
Filter BAM or SAM on data 140.bam.nsort.bam.fixed.bam
Filter BAM or SAM on data 188.bam.nsort.bam.fixed.bam
Filter BAM or SAM on data 236.bam.nsort.bam.fixed.bam
Filter BAM or SAM on data 92.bam.nsort.bam.fixed.bam
>foreach f ( *.filt.bam )
samtools sort "$f" > "$f.csort.bam"
echo $f
end
```

Not sure why the workflow chokes on samtools sort.

**Comment by hepcat72** [comment]
I was going through the history to clear out the stuff I ended up doing on the command line.  It's notable that each sort job has a dataset that succeeded and one that fails (both with the same number):

![sortsucceeds_and_fails](https://user-images.githubusercontent.com/2300532/65721273-baba1c80-e077-11e9-9318-bdc05babd663.png)

**Comment by mvdbeek** [comment]
You should see an error in the logs indicating why the job couldn't be finished. That said you've sorted the bam file by name, and then fed it into a tool that uses coordinate sorted files, which triggers the auto-conversion to coordinate sorted files. This also doesn't seem right.

Given that both samtools sort and implicit conversions work fine on usegalaxy.org and on my test instance I'm going to close this, but feel free to re-open if you have any logs indicating what is going wrong.

**Comment by hepcat72** [comment]
The tool I feed the name-sorted bam file to is `samtools fixmate`, which requires a name-sorted bam. Its the end of this path:

paired collection->cutadapt->RNA STAR->RmDup->Filter BAM or SAM->samtools sort...

which goes into fixmate.  It's fixmate and all subsequent steps which get paused.

Am I wrong?  Does samtools fixmate require coordinate-sorted data?  The only other samtools sort I do in the flow is not outputing to any other tool.  So I'm confused by your assessment...

https://www.dropbox.com/s/do83iec02aks8ub/namesorttofixmate.png?dl=0

(I can't seem to upload an image at the moment.)

But I'll see if I can find some time to run the toy example again and look for the error in the log.

**Comment by mvdbeek** [comment]
https://github.com/galaxyproject/tools-iuc/blob/master/tool_collections/samtools/samtools_fixmate/samtools_fixmate.xml#L34 this means input is coordinate sorted. samtools fixmate itself works on named sorted files, but not the galaxy wrapper. there's a sort step in the wrapper itself.

This is something to improve in the wrapper for sure, but given that the implicit conversion from name sorted to coordinate sorted works in other galaxy servers I don't think this is a bug in Galaxy.

**Comment by hepcat72** [comment]
Ah, I see.  It seems I have a hard time thinking of these tools in a way other than in the command-line context.  Unlearn what I have learned, I must.

**Comment by hepcat72** [comment]
I was telling Lance about this at lunch.  He said that you will *always* get an error from samtools sort if you connect the output to something else because it will fail when it automatically tries to index it.  You can't index a name-sorted bam file.  And he said since it's not a tool error, you won't see a reason for the failure in your history.

Lance said that he felt that either there should not be a name-sort tool in galaxy or galaxy's automatic format conversion should be smarter about this.  We're a little behind, version-wise, so perhaps this has already been addressed - since you said it works on usegalaxy.org?  If not, perhaps I should create an issue to address this?

**Comment by mvdbeek** [comment]
This isn't specific enough to work on. Barring any bugs (those are always on the table) everything should work correctly in the latest versions of everything. Name-sorting tools will produce `qname_sorted.bam`, which are not going to be indexed.

**Comment by hepcat72** [comment]
OK. In the future, I will try to test issues I run into on usegalaxy before submitting.

**Comment by hepcat72** [comment]
So, to follow through on this, I just tested this on our galaxy with a small dataset and reproduced the issue.  Note, you have to view hidden files to see that there was a failure:

<img width="285" alt="samtools_sort_failure" src="https://user-images.githubusercontent.com/2300532/67719434-30334880-f9a8-11e9-993e-e8850874088e.png">

The error in the log was:

```
galaxy.jobs DEBUG 2019-10-28 17:12:51,322 (252766) Persisting job destination (destination id: cetus_singlecore)
galaxy.jobs.runners.drmaa DEBUG 2019-10-28 17:12:51,489 (252766/30689804) state change: job is queued and active
galaxy.jobs.runners.drmaa DEBUG 2019-10-28 17:12:52,501 (252766/30689804) state change: job is running
galaxy.jobs.runners.drmaa DEBUG 2019-10-28 17:12:57,584 (252766/30689804) state change: job finished normally
galaxy.jobs.output_checker DEBUG 2019-10-28 17:12:57,622 job failed, standard error is - [sort: invalid option -- '@'
sort: invalid option -- 'O'
sort: invalid option -- 'T'
open: No such file or directory
[bam_sort_core] fail to open file 1
]
galaxy.model.metadata DEBUG 2019-10-28 17:12:57,663 setting metadata externally failed for HistoryDatasetAssociation 583588: 'samtools returned with error 1: stdout=, stderr=samtools index: "/galaxy/files/000/474/dataset_474749.dat" is in a format that cannot be usefully indexed\n'
galaxy.jobs.runners ERROR 2019-10-28 17:12:57,677 (252766/30689804) Job wrapper finish method failed
Traceback (most recent call last):
  File "/galaxy/galaxy-app/lib/galaxy/jobs/runners/__init__.py", line 717, in finish_job
    self._finish_or_resubmit_job(job_state, stdout, stderr, exit_code)
  File "/galaxy/galaxy-app/lib/galaxy/jobs/runners/__init__.py", line 459, in _finish_or_resubmit_job
    job_state.job_wrapper.finish(stdout, stderr, exit_code, check_output_detected_state=check_output_detected_state)
  File "/galaxy/galaxy-app/lib/galaxy/jobs/__init__.py", line 1371, in finish
    dataset.datatype.set_meta(dataset, overwrite=False)
  File "/galaxy/galaxy-app/lib/galaxy/datatypes/binary.py", line 432, in set_meta
    pysam.index(dataset.file_name, index_file.file_name)
  File "/galaxy/galaxy-app/.venv/lib/python2.7/site-packages/pysam/utils.py", line 75, in __call__
    stderr))
SamtoolsError: 'samtools returned with error 1: stdout=, stderr=samtools index: "/galaxy/files/000/474/dataset_474749.dat" is in a format that cannot be usefully indexed\n'
galaxy.jobs DEBUG 2019-10-28 17:12:57,745 Pausing Job '252767', Execution of this dataset's job is paused because its input datasets are in an error state.
galaxy.tools.error_reports DEBUG 2019-10-28 17:12:57,774 Bug report plugin <galaxy.tools.error_reports.plugins.sentry.SentryPlugin object at 0x7f9f4f496d50> generated response None
galaxy.model.metadata DEBUG 2019-10-28 17:12:57,775 Cleaning up external metadata files
galaxy.model.metadata DEBUG 2019-10-28 17:12:57,780 Failed to cleanup MetadataTempFile temp files from /galaxy/galaxy-app/database/jobs_directory/000/252/252766/metadata_out_HistoryDatasetAssociation_583588_kxRquz: No JSON object could be decoded
```

which is what Lance predicted would happen.  It tried to index the name-sorted bam and failed and the samtools sort step got paused.

I also uploaded the toy workflow and the test data to usegalaxy.org and started a test there.  It's queued up and is just waiting to be run.  I was going to wait for it before posting this, but it's taking awhile, so I'm just going to send this now and link you to the usegalaxy history to see if the hidden failed dataset shows up there too.

https://usegalaxy.org/u/hepcat_72/h/unnamed-history

Hopefully, it succeeds and this issue is specific to our version of galaxy!

**Comment by mvdbeek** [comment]
```
galaxy.jobs.output_checker DEBUG 2019-10-28 17:12:57,622 job failed, standard error is - [sort: invalid option -- '@'
sort: invalid option -- 'O'
sort: invalid option -- 'T'
open: No such file or directory
[bam_sort_core] fail to open file 1
]
```

you're not using the right samtools version. Please make sure you have all dependencies correctly installed in the Manage dependencies view in the admin panel.
For your case in particular the tools to check the dependencies for are
```
CONVERTER_bam_to_qname_sorted_bam
CONVERTER_bam_to_coodinate_sorted_bam
CONVERTER_sam_to_unsorted_bam
```

**Comment by hepcat72** [comment]
Yeah, they all had a red triangle with an exclamation point whose tooltip says "Dependency resolved but version 0.1.18 not found".  The resolver in each case was "Galaxy_Package".  Does the "dependency resolved" message mean only that *a* version of the tool was found, or does it also mean that the version found is *less* than the version needed, because in each case, the "Exact" column shows "False".

Regardless, I installed them using conda and now they all say version 1.6 and resolver is conda.  I re-ran the toy example and there is not longer an error.
